lib/resource-timing.js

```diff
@@ -0,0 +1,228 @@
+   *
+   * @class SurfNPerf
+   */
+  var SurfNPerfRT = function() {
+      this._data = {
+        custom: {},
+        marks: {},
+        highResMarks: {},
+        events: {}
+      };
+
+      this._navigationTiming = null;
+      this._highResTime = null;
+      this._userTiming = null;
+
+      this._navigationTimingEvents = {
+        a: ["navigationStart", "unloadEventEnd", "unloadEventStart", "redirectStart", "redirectEnd", "fetchStart", "domainLookupStart", "domainLookupEnd", "connectStart", "secureConnectionStart", "connectEnd", "requestStart", "responseStart", "responseEnd", "domLoading"],
+        b: ["domInteractive", "domContentLoadedEventStart", "domContentLoadedEventEnd", "domComplete", "loadEventStart", "loadEventEnd"]
+      };
+
+      this.initialize();
+    },
+    SNPRTProto = SurfNPerfRT.prototype;
+
+  SNPRTProto._inList = function(origin, options) {
+    if (options.hasOwnProperty("whitelist")) {
+      if (options.whitelist.indexOf(origin) === -1) {
+        return false;
+      } else {
+        return true;
+      }
+    } else if (options.hasOwnProperty("blacklist")) {
+      if (options.blacklist.indexOf(origin) === -1) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  SNPRTProto.getOrigins = function(options) {
+    if(window.performance.getEntriesByType) {
+      var resources = window.performance.getEntriesByType("resource");
+      var origins = [];
+      for var resource in resources {
+        var url = new URL(resources[resource].name);
+        if (origins.indexOf(url.origin) === -1 && this._inList(url.origin, options)) {
+          origins.push(url.origin);
+        }
+      }
+
+      return origins;
+    } else {
+      return null;
+    }
+  };
+
+  SNPRTProto.getResourcesFromOrigin = function(origin) {
+    if(window.performance.getEntriesByType) {
+      var resources = window.performance.getEntriesByType("resource");
+      var resourcesFromOrigin = [];
+      for var resource in resources {
+        var url = new URL(resources[resource].name);
+        if (url.origin === origin) {
+          resourcesFromOrigin.push(resources[resource]);
+        }
+      }
+      return resourcesFromOrigin;
+    } else {
+      return null;
+    }
+  };
+
+  SNPRTProto.isBufferFull = function() {
+    if(window.performance.getEntriesByType) {
+      if(performance.onresourcetimingbufferfull) {
+        return true;
+      } else {
+        return false;
+      }
+    } else {
+      return null;
+    }
+  };
+
+  SNPRTProto._nameExtend = function(name) {
+    if(name.charAt(0) == '/') {
+      return "http://" + window.location.hostname + name;
+    } else {
+      return name;
+    }
+  };
+
+  SNPRTProto.getResource = function(name) {
+    name = this._nameExtend(name);
+
+    return window.performance.getEntriesByName(name, "resource")[0]
+  };
+
+  SNPRTProto.duration = function(name, eventA, eventB, options) {
+    name = this._nameExtend(name);
```

Comment:
you don't have to do `this._name(name)` here since `getResource` takes care of that for you